Automattic/wp-calypso
# Move and modularise payment state

This PR is one of many working on modularising state in Calypso. This one handles payment state.

For more details on state modularisation, see the [modularised state documentation](https://github.com/Automattic/wp-calypso/blob/master/docs/modularized-state.md) and p4TIVU-9lM-p2.

Note: I added all the reviewers that GitHub suggested. Please feel free to ignore the review request or pull in someone else if you're not comfortable reviewing this PR. Thank you!

#### Changes proposed in this Pull Request

* Move `payment` state out of `ui` and into a new top level entry
* Modularise it
* Extract shared function to a util file.

#### Testing instructions

* Install [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en).
* Open DevTools and switch to the Redux tab.
* Prepare a non-empty cart on the same account you'll be testing with.
* Open `/home/<your-site>` on the live branch or your local copy of this branch.
* Ensure that the `Your cart is awaiting payment` notification is shown.
* Click `State` on the top right to ensure you see all state, and not just the diff for the last action: 
![image](https://user-images.githubusercontent.com/409615/73774406-d3e87d80-477b-11ea-9f59-3e42cc00101f.png)
* Analyse the tree and note that there's no `payment` key, even if you expand the list of keys. This indicates that the payment state hasn't been loaded yet.
* Click `View your cart` in the `Your cart is awaiting payment` notification.
* Verify that a `payment` key is added with the payment state. This indicates that the payment state has now been loaded.
* Verify that payment-related functionality works normally, by using the feature.
